allennlp/service/config_explorer.html

```diff
@@ -0,0 +1,608 @@
+                              .filter(x => x !== undefined)
+                              .toArray()
+
+            return list.length ? list : undefined
+        } else if (origin === 'int' || origin === 'float') {
+            const numeric = +value
+            return value.length && !isNaN(numeric) ? numeric : undefined
+        } else if (origin === 'bool') {
+            return {'true': true, 'false': false}[value]
+        } else if (origin === 'str') {
+            return value.length ? value : undefined
+        } else {
+            console.log("unknown type " + annotation.toJS())
+            return undefined
+        }
+    }
+}
+
+const configToJson = (config) => {
+
+    let blob = {}
+
+    if (config) {
+        const type = config.get('type')
+        if (type) {
+            blob['type'] = type
+        }
+
+        config.get('items').forEach((item) => {
+            const value = item.get('value')
+            const annotation = item.get('annotation')
+            const configurable = item.get('configurable')
+
+            const json = jsonify(value, annotation, configurable)
+
+            if (json !== undefined) {
+                const name = item.get('name')
+                blob[name] = json
+            }
+        })
+    }
+
+    return blob
+}
+
+// Each node in the config tree will be marked complete if either
+// * it's optional, or
+// * it's a leaf and has a value, or
+// * all of its children are complete
+//
+// This means it's a bottom up concept. This is the one place
+// where immutability makes our lives a little trickier.
+const markComplete = (node) => {
+    const optional = node.get('defaultValue') !== undefined
+    const origin = node.getIn(['annotation', 'origin'])
+    const value = node.get('value')
+
+    if (node.get('configurable')) {
+        if (value) {
+            const items = node.getIn(['value', 'items']).map(markComplete)
+            return node.set('completed', optional || items.every(item => item.get('completed')))
+                    .setIn(['value', 'items'], items)
+        } else {
+            return node.set('completed', optional)
+        }
+    } else if (Immutable.List.isList(value)) {
+        const items = value.map(markComplete)
+        return node.set('completed', optional || items.every(item => item.get('completed')))
+                   .set('value', items)
+    } else {
+        const completed = optional || value
+        return node.set('completed', completed)
+    }
+}
+
+// Top-level component
+class App extends React.Component {
+    constructor() {
+        super()
+
+        this.state = {
+            data: Immutable.Map()
+        }
+
+        this.setData = this.setData.bind(this)
+    }
+
+    setData(fn) {
+        const {data} = this.state
+        const newData = markComplete(fn(data))
+        return this.setState({data: newData})
+    }
+
+    componentDidMount() {
+        // Fetch the top level configuration
+        fetch('/api/')
```

Comment:
Is it possible that this API is going to be extended to include partial json configurations/recommendations etc? It might be better to call this api `/annotate` or something, if that is the case.

Reply:
I'll make it `/api/config/` to leave room for other `/api/` routes